src/services/upower/core.hpp

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <qdbusservicewatcher.h>
+#include <qhash.h>
+#include <qobject.h>
+#include <qqmlintegration.h>
+#include <qqmllist.h>
+#include <qtmetamacros.h>
+
+#include "../../core/model.hpp"
+#include "dbus_service.h"
+#include "device.hpp"
+
+namespace qs::service::upower {
+
+class UPower: public QObject {
+	Q_OBJECT;
+
+public:
+	[[nodiscard]] UPowerDevice* displayDevice();
+	[[nodiscard]] ObjectModel<UPowerDevice>* devices();
+
+	static UPower* instance();
+
+signals:
+	void displayDeviceChanged();
+
+private slots:
+	void onDeviceReady();
+	void onDeviceDestroyed(QObject* object);
+
+private:
+	explicit UPower();
+
+	void registerExisting();
+	void registerDevice(const QString& path);
+
+	UPowerDevice* mDisplayDevice = nullptr;
+	QHash<QString, UPowerDevice*> mDevices;
+	ObjectModel<UPowerDevice> readyDevices {this};
+
+	DBusUPowerService* service = nullptr;
+};
+
+class UPowerQml: public QObject {
+	Q_OBJECT;
+	QML_NAMED_ELEMENT(UPower);
+	QML_SINGLETON;
+	/// UPower's DisplayDevice for your system.
+	///
+	/// This is typically the device that is used for displaying information in desktop environments. Can be `null`.
+	Q_PROPERTY(UPowerDevice* displayDevice READ displayDevice NOTIFY displayDeviceChanged);
```

Comment:
Mention that this is an aggregate device and not a physical device, and therefore is not present in the device list.